app/controllers/friend_requests_controller.rb

```diff
@@ -0,0 +1,47 @@
+# require 'byebug'
+
+class FriendRequestsController < ApplicationController
+  before_action :set_receiver_id, except: :index
+  before_action :validate_creation, only: :create
+  before_action :validate_inverse_request, only: :create
+
+  def index
+    @requests = current_user.friend_requests
+  end
+
+  def create
+    @friend_request = current_user.send_friend_request(@receiver_id)
+    redirect_back fallback_location: users_path
+  end
+
+  def destroy
+    current_user.delete_friend_request(@receiver_id)
+    redirect_back fallback_location: friend_requests_path
+  end
+
+  private
+
+  def friend_request_params
+    params.require(:friend_request).permit(:receiver_id)
+  end
+
+  def set_receiver_id
+    @receiver_id = friend_request_params[:receiver_id].to_i
+  end
+
+  def validate_creation
+    return unless current_user.friend?(@receiver_id) || current_user.friend_requested?(@receiver_id)
+
+    flash[:alert] = 'This user is already your friend or the friend request have been done before'
+    redirect_back fallback_location: users_path
+  end
+
+  def validate_inverse_request
```

Comment:
Our friendship creation in relation to friend_request is a little too complex. We don't need special methods such as these to create a friendship and then delete the associated friend_request:

```ruby
#friendships_controller.rb

def create
 @friendship = Friendship.new(friendship_params)

 if @friendship.save
   # do something
 else
  # do something else
 end
end
```

Likewise, we can have simple `create` and `destroy` methods in our `friend_requests_controller.rb`:

```ruby
#friend_requests_controller.rb

def create
 # we can simply call `new` and/or `save` class methods for `FriendRequest` model here
 # no need for special instance methods for current_user
end

def destroy
  # we can simply find the FriendRequest record we wanna delete here by using
 # FriendRequest.find or FriendRequest.find_by and then call on that the .delete method 
# we don't need to have a special delete_request method for current_user
end
```

To facilitate the deletion of associated request, we can simply have that **special** method in the `Friendship` model that deletes the request `after_create`/`after_save`:

```ruby
# friendship.rb

after_save :delete_this_request

def delete_this_request(# necessary arguments)
  # when Paul accepts Mark's friend request delete the request here
end
```